### DISCUSSION TITLE
Pulsed Confocal Scan

### DISCUSSION THREAD

**Discussion by Fluxembourg** [Ulm-IQO/qudi-core#54] [Q&A]
Hi all,

I am quite new to programming/Qudi, but I would like to develop an "improved" version of the standard confocal scan.

**Basic idea:**
Use PulseStreamer as clock signal distributer.
Use TimeTagger for counting photons.
Use NIDAQ card for analog voltages (to the scanner).

**In my understanding this would involve:**
1.) Definining a list of voltage sets for the NIDAQ card (xy scan)
2.) Triggering the NIDAQ card every time that a new set of voltage values should be set.
3.) Triggering the TimeTagger to count photons during the integration interval.

In my naive understanding, I could do this by providing the clock signal of the confocal scan via the PulseStreamer.
However, I am a bit lost as to where to start implementing these changes.

Thanks for any help.

**Comment by LukePiewalker** [comment]
Dear @Fluxembourg,

I guess the starting point for you would be to implement the functionality to provide an external sample clock to the `NIXSeriesFiniteSamplingIO`. Potentially you could use the `NIXSeriesInStreamer` as a guide, as there is some implementation of using an external sample clock, however so far not really tested to my knowledge. Correct me if I'm wrong @Neverhorst?

Second step one would need is somehow getting back the signal from the TimeTagger in some way to make it accessible with the scanning probe toolchain. Depending on your exact use case, I have different options in mind. Would you also like to have the pulsed/gated signal in the end on a slow counter while dragging the cursor in your scan etc?

**Comment by timoML** [comment]
In addition to that, a short clearification on how to use interfuses that was raised in private communication:
- On the measurement (in qudi terms "logic") level, it is easy to control multiple hardwares by simply using the Connector infrastructure.
- If you're writing a hardware module that should fulfill a certain interface (eg. here the `ScanningProbeInterface`), interfuses are the way to combine multiple different hardwares. For the NI hardware for scanning, we do this in `NiScanningProbeInterfuse`: Here the ni_ao module is used to give control voltages to the scanner and the ni_finite_sampling_io module is reading counts from the nicard. In your case, you indeed could write a similar interfuse that includes the TimeTagger. As @LukePiewalker mentioned above, another way would be to extend the TimeTagger to fullfill the `FiniteSamplingIOInterface`. Then, it should be fairly simple to just employ the TimeTagger as FiniteSampling device in the existing `NiScanningProbeInterfuse`.
